### PR TITLE
support sharding for probe

### PR DESCRIPTION
**Pull request by steven05jiang**
## Description

fix https://github.com/prometheus-operator/prometheus-operator/issues/5176. 
By adding a sharding relabeling rule, prometheus operator will shard targets in the probe across the prometheus cluster.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
support sharding for probe.
```

### CODE REVIEW

**Review by soasme — COMMENTED** [review]
The fix looks good to me.

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -957,6 +957,7 @@ func (cg *ConfigGenerator) generateProbeConfig(
 		// Add configured relabelings.
 		xc := labeler.GetRelabelingConfigs(m.TypeMeta, m.ObjectMeta, m.Spec.Targets.StaticConfig.RelabelConfigs)
 		relabelings = append(relabelings, generateRelabelConfig(xc)...)
+		relabelings = generateAddressShardingRelabelingRulesForProbes(relabelings, shards)
```
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:960]: Do we want to configure sharing on every occasion? Or only when sharding is enabled in Prometheus?

**Review by ArthurSens — COMMENTED** [review]
Hello 👋 , sorry for taking this long to review... this one is a bit out of my comfort zone 😅

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (steven05jiang) [reply pkg/prometheus/promcfg.go]: thanks for the suggestion. this is the change [e4ded8d](https://github.com/prometheus-operator/prometheus-operator/pull/5735/commits/e4ded8d7fd91d4d9874cb87a4cb88c1e31f1b2a2)

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -909,6 +909,7 @@ func (cg *ConfigGenerator) generateProbeConfig(
 			},
 		}...)
 	}
+	relabelings = generateAddressShardingRelabelingRulesForProbes(relabelings, shards)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:912]: I'd keep it as the last relabeling config.

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:912]: @steven05jiang yes I think that it makes sense to do so.

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -1053,10 +1052,11 @@ func (cg *ConfigGenerator) generateProbeConfig(
 
 		// Add configured relabelings.
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(m.TypeMeta, m.ObjectMeta, m.Spec.Targets.Ingress.RelabelConfigs))...)
-		relabelings = generateAddressShardingRelabelingRulesForProbes(relabelings, shards)
+	}
 
+	if m.Spec.Targets.StaticConfig != nil || m.Spec.Targets.Ingress != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1057]: I don't think that the if statement is required
```suggestion
```

**Review by simonpasquier — COMMENTED** [review]
one small nit otherwise lgtm!

File: pkg/prometheus/promcfg.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:1057]: This is a good remark. The operator already takes care of filtering out probes without static/ingress targets [here](https://github.com/prometheus-operator/prometheus-operator/blob/4e37c62781ce19971726f5a34b073327a3860d1a/pkg/prometheus/resource_selector.go#L497-L500). But even if the check doesn't work, it shouldn't be a problem to include the relabeling configs.

Comment (steven05jiang) [reply pkg/prometheus/promcfg.go]: Sounds good. This is the update ffb2a35

**Review by ArthurSens — APPROVED** [review]
good job @steven05jiang and @simonpasquier!

sorry for not being very useful here 😅

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Comment by steven05jiang** [comment]
Thanks all for the review!

**Pull request merged**